### PR TITLE
Re-write error handling

### PR DESCRIPTION
**Pull request by milankinen**
This PR moves individual error handling from different namespaces (e.g `consumer`, `publish`) to single place: `channel` namespace . Now the channel namespace provides only **one** "public" function for other namespaces which is `with-channel`. In order to use this function, callers must implement `channel.Runner` protocol and assign error handling semantics for each error types.

For simplicity, error types are reduced into three categories:
* Connection errors - errors that are neither client's nor broker's fault, hence recoverable after the connection restores
* Protocol errors - errors that are caused (usually) by client, e.g. trying to publish to non-existing exchange, consume exclusively consumed queue etc... Depending on the error type and situation, recovering from these errors might be possible or not
* Throwables - all other types of errors, caused by programmer/library bugs. Recovering depends on the usage

Interrupted errors are propagated as they are and `with-channel` users are responsible for dealing with them properly.

In order to react&handle different kind of errors cleanly, we can't use auto-recovering channels internally (their internal register in RabbitMQ client messes up things and cause unnecessary complexity on connection recovery situations). However, the automatic connection recovery is needed. The existing `AutorecoveringConnection` implementation does not provide any way to create "non-recovering" channels, hence this PR introduces an own, simplified, implementation for auto-recovery connections, based on `com.rabbitmq.client.AutoRecoveringConnection`. 

**NOTE FOR REVIEWERS**: this PR re-writes basically whole codebase so it might be easier to review changes by looking the re-written sources, starting from `internal.channel` namespace and expanding review to other internal modules in natural dependency order and finally reviewing the public core namespace (and tests). Reviewing diffs probably doesn't make any sense.

#ovp-board

### CODE REVIEW

File: src/clj/rabid/core.clj [hunk 1]
```diff
@@ -44,17 +48,17 @@
         (reset! ctx' nil)))))
 
 (defn consume!
-  ([queue handler exclusive?]
+  ([^String queue ^IFn handler ^Boolean exclusive?]
    (locking ctx-lock
      (let [{:keys [opts] :as ctx} (get-ctx)
            intervals {:reconnect  (or (:consumer-reconnect (:intervals opts)) 10000)
                       :ha-polling (or (:exclusive-consumer-poll (:intervals opts)) 10000)}
```
Comment (mjokinen-affecto) [inline src/clj/rabid/core.clj]: Defaultit vois ehkä laittaa `def`:iksi/:eiksi

File: src/clj/rabid/internal/channel.clj [hunk 2]
```diff
@@ -1,76 +1,118 @@
...
+    (handler-fn ch))
+  (on-connection-error [this e]
+    (on-throwable this e))
+  (on-protocol-error [this e]
+    (on-throwable this e))
+  (on-throwable [_ e]
+    (throw e)))
+
+(defn on-shutdown [ch callback]
+  (let [callback' (once callback)]
```
Comment (mjokinen-affecto) [inline src/clj/rabid/internal/channel.clj]: Voisko tämän inlineata?

Comment (milankinen) [inline src/clj/rabid/internal/channel.clj]: Ei voi koska muuten luotaisiin once joka kerta uusiksi kun shutdown-listeneriä kutsutaan

**Comment by mjokinen-affecto** [comment]
Hyvältä näyttää!

**Pull request merged**